app.py

```diff
@@ -0,0 +1,283 @@
+)
+
+class CoOp(db.Model):
+  id = db.Column(db.Integer(), primary_key=True)
+  name = db.Column(db.String(80), unique=True)
+  is_active = db.Column(db.Boolean())
+  start_date = db.Column(db.DateTime())
+  end_date = db.Column(db.DateTime())
+  location = db.Column(db.String(255))
+  interest = db.Column(db.Integer())
+  initial_balance = db.Column(db.Integer())
+  expected_repayment = db.Column(db.Integer())
+  current_balance = db.Column(db.Integer())
+
+  def __str__(self):
+    return self.name
+
+
+class Role(db.Model):
+  id = db.Column(db.Integer(), primary_key=True)
+  name = db.Column(db.String(80), unique=True)
+  description = db.Column(db.String(255))
+
+  def __str__(self):
+    return self.name
+
+
+class User(db.Model):
+  id = db.Column(db.Integer, primary_key=True)
+  first_name = db.Column(db.String(255))
+  last_name = db.Column(db.String(255))
+  email = db.Column(db.String(255), unique=True)
+  phone = db.Column(db.String(255), unique=True)
+  password = db.Column(db.String(255))
+  active = db.Column(db.Boolean())
+  confirmed_at = db.Column(db.DateTime())
+  loan = db.relationship('Loan', uselist=False, backref='user')
+  co_ops = db.relationship('CoOp', secondary=co_ops_users,
+                          backref=db.backref('users', lazy='dynamic'))
+  roles = db.relationship('Role', secondary=roles_users,
+                          backref=db.backref('users', lazy='dynamic'))
+
+  def __str__(self):
+    return self.email
+
+class Loan(db.Model):
+  id = db.Column(db.Integer, primary_key=True)
+  balance = db.Column(db.Integer)
+  interest = db.Column(db.Integer)
+  loan_start = db.Column(db.DateTime)
+  loan_end = db.Column(db.DateTime)
+  user_id = db.Column(db.Integer, db.ForeignKey('user.id'))
+
+class Transaction(db.Model):
+  id = db.Column(db.Integer, primary_key=True)
+  amount = db.Column(db.Integer)
+  previous_balance = db.Column(db.Integer)
+  new_balance = db.Column(db.Integer)
+  loan = db.relationship('Loan', secondary=transactions_loans,
+                          backref=db.backref('transactions', lazy='dynamic'))
+  user = db.relationship('User', secondary=transactions_users,
+                          backref=db.backref('transactions', lazy='dynamic'))
+
+# Customized admin interface
+class CustomView(ModelView):
+  list_template = 'list.html'
+  create_template = 'create.html'
+  edit_template = 'edit.html'
+
+
+class UserAdmin(CustomView):
+  column_searchable_list = ()
+  column_filters = ('first_name', 'email')
+  column_exclude_list = ['password', ]
+
+
+# Flask views
+@app.route('/', methods = ['GET', 'POST'])
+def index():
+  if request.method == 'GET':
+    return '<a href="/admin/">Click me to get to Admin!</a>'
+  if request.method == 'POST':
+    msg = request.form.getlist('text')
+    cmd = msg[0].split()[0]
+    if cmd.lower() == 'loan':
+      add_transaction(request.form)
+      return 'success: added loan transaction'
+    # ImmutableMultiDict([
```

Comment:
What is this section for?

Reply:
ah I just left that there to remind myself what the incoming messages look like to set up the parsing - I'll probably be removing them once I have the rest of that set up